feature/experimental/p4rt/ate_tests/performance_test/performance_test.go

```diff
@@ -0,0 +1,932 @@
+			pktIn := 0
+			// Run Packetin and packetout traffic in parallel.
+			var wg sync.WaitGroup
+			wg.Add(4)
+
+			go func() {
+				defer wg.Done()
+				pktIn = testTraffic(t, args.ate, flows, srcEndPoint, 20)
+				t.Logf("Total Packetin packets sent from ATE %v", pktIn)
+			}()
+
+			go func() {
+				defer wg.Done()
+				time.Sleep(packetOutWait)
+				sendPackets(t, test.client, packets[0], packetCount, 2.6) // GDP packetout with 2.6ms timer.
+			}()
+
+			go func() {
+				defer wg.Done()
+				time.Sleep(packetOutWait)
+				sendPackets(t, test.client, packets[1], packetCount, 3.1) // Traceroute packetout with 3.1ms timer.
+			}()
+
+			go func() {
+				defer wg.Done()
+				time.Sleep((packetOutWait))
+				sendPackets(t, test.client, packets[2], packetCount, 5.1) // LLDP packetout with 5.1ms timer.
+			}()
+
+			wg.Wait() // Wait for all four goroutines to finish before exiting.
+
+			test.wantPkts = pktIn
+			// Check packet counters after packet out
+			counter1 := gnmi.Get(t, args.ate, gnmi.OC().Interface(port).Counters().InPkts().State())
+
+			// Verify Packetout stats to check P4RT stream
+			t.Logf("Received %v packetout on ATE port %s", counter1-counter0, port)
+
+			if test.expectPass {
+				if (counter1 - counter0) < uint64(float32(3*packetCount)*0.95) {
+					t.Fatalf("Number of Packetout packets, got: %d, want: %d", counter1-counter0, (3 * packetCount))
+				}
+			} else {
+				if (counter1 - counter0) > uint64(float32(3*packetCount)*0.10) {
+					t.Fatalf("Unexpected packets are received.")
+				}
+			}
+			_, packetin_pkts, err := test.client.StreamChannelGetPackets(&streamName, uint64(test.wantPkts), 30*time.Second)
+			if err != nil {
+				t.Errorf("Unexpected error on StreamChannelGetPackets: %v", err)
+			}
+
+			if got, want := len(packetin_pkts), test.wantPkts; got != want {
+				t.Errorf("Number of PacketIn, got: %d, want: %d", got, want)
+			}
+			if test.wantPkts == 0 {
+				return
+			}
+			t.Logf("Start to decode packetin and compare with expected packets.")
+			wantgdpPacket := args.gdpPacketIO.GetPacketTemplate()
+			wantlldpPacket := args.lldpPacketIO.GetPacketTemplate()
+			wanttrPacket := args.trPacketIO.GetPacketTemplate()
+
+			gdpIncount := 0
+			lldpIncount := 0
+			trIncount := 0
+
+			for _, packet := range packetin_pkts {
+				if packet != nil {
+					dstMac, etherType := decodeL2Packet(packet.Pkt.GetPayload())
+					ttl := decodeIPPacketTTL(packet.Pkt.GetPayload())
+					metaData := packet.Pkt.GetMetadata()
+					if wantgdpPacket.EthernetType != nil {
+						if etherType == layers.EthernetType(*wantgdpPacket.EthernetType) {
+							if dstMac == *wantgdpPacket.DstMAC {
+								for _, data := range metaData {
+									switch data.GetMetadataId() {
+									case metadataIngressPort:
+										if string(data.GetValue()) != args.gdpPacketIO.GetIngressPort() {
+											t.Fatalf("Ingress Port Id is not matching expectation for GDP.")
+										}
+									case metadataEgressPort:
+										found := false
```

Comment:
This can be more concise and look more like the ingress port case:

```
if string(data.GetValue()) != args.gdpPacketIO.GetEgressPort() {
	t.Fatalf("Egress Port Id is not matching expectation for GDP.")
}
gdpIncount += 1
```